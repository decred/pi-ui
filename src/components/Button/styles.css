.btn,
.btn:link,
.btn:visited {
  font-size: var(--font-size-normal);
  line-height: var(--spacing-2);
  padding: 1rem 2.5rem;
  cursor: pointer;
  text-decoration: none;
  border-radius: 0.5rem;
  transition: all 0.2s;
  backface-visibility: hidden;
}

.btn:not(:last-child):not(:only-child) {
  margin-right: var(--spacing-1);
}

.primary {
  composes: btn;
  color: var(--color-white);
  background-color: var(--btn-color);
  border: 1px solid var(--btn-color);
  box-shadow: 1px 1px 5px 1px var(--color-shadow-light);
}

.secondary {
  composes: btn;
  color: var(--btn-color);
  background-color: transparent;
  border: 1px solid var(--btn-color);
  box-shadow: 1px 1px 5px 1px var(--color-shadow-light);
}

.disabled {
  composes: btn;
  cursor: not-allowed;
  color: var(--color-gray);
  background-color: var(--color-gray-lighter);
  border: 1px solid var(--color-gray-lighter);
}

.primary:focus,
.primary:hover,
.secondary:focus,
.secondary:hover {
  outline: 0;
  transform: scale(1.01);
  box-shadow: 1px 1px 10px 3px var(--color-shadow-light);
}

.primary:active,
.secondary:active {
  transform: scale(1);
  box-shadow: 0 0 0;
}

.fullWidth {
  width: 100%;
}
