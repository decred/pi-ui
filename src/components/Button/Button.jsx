import PropTypes from "prop-types";
import React from "react";
import { classNames } from "../../utils";
import Spinner from "../Spinner/Spinner.jsx";
import styles from "./styles.css";

const Button = ({
  children,
  className,
  style,
  kind,
  size,
  icon,
  onClick,
  loading,
  fullWidth,
  noBorder,
  width,
  type,
  ...props
}) => (
  <button
    type={type}
    className={classNames(
      styles[kind],
      styles[size],
      icon && styles.icon,
      fullWidth && styles.fullWidth,
      noBorder && styles.noBorder,
      className
    )}
    style={width ? { width: `${width}px`, ...style } : style}
    disabled={kind === "disabled" || loading}
    onClick={onClick}
    {...props}>
    {loading ? <Spinner /> : children}
  </button>
);

Button.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  style: PropTypes.object,
  onClick: PropTypes.func,
  loading: PropTypes.bool,
  fullWidth: PropTypes.bool,
  icon: PropTypes.bool,
  width: PropTypes.number,
  size: PropTypes.oneOf(["sm", "md", "lg"]),
  type: PropTypes.oneOf(["button", "submit"]),
  kind: PropTypes.oneOf(["primary", "secondary", "disabled"]),
  noBorder: PropTypes.bool,
};

Button.defaultProps = {
  kind: "primary",
  size: "md",
  type: "button",
  icon: false,
  loading: false,
  fullWidth: false,
  noBorder: false,
};

export default Button;
