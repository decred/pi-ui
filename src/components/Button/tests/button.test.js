import React from "react";
import Button from "../Button";
import { create } from "react-test-renderer";
import { fireEvent, render } from "@testing-library/react";

describe("Button component", () => {
  test("Matches the snapshot", () => {
    const button = create(<Button>test</Button>);
    expect(button.toJSON()).toMatchSnapshot();
  });

  test("Calls onSubmit() prop on submit button click", () => {
    const mockedOnClick = jest.fn();

    const { getByText, queryByText } = render(
      <Button onClick={mockedOnClick}>test</Button>
    );

    // button rendered with right children
    expect(queryByText("test")).toBeTruthy();

    // click button
    fireEvent.click(getByText("test"));

    // expect click cb to be called
    expect(mockedOnClick).toHaveBeenCalledTimes(1);
  });
});
