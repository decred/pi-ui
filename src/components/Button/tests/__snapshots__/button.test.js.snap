// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Matches the snapshot 1`] = `
<button
  className="primary md"
  disabled={false}
  type="button"
>
  test
</button>
`;
